docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+| **NewVsix** | The new VS Extension which will be powered by LSP | with VS |
+
+
+## Components
+
+### FsLSP
+
+The library containing the LSP functionality. Can be used to create an LSP server by adding the transport layer.
+
+#### FsLSPServer
+
+The LSP server executable using STD IN/OUT for transport.
+
+Will be packaged with FsLSP and available as a dotnet tool (?)
+
+It should always target the latest .NET runtime.
+
+#### Workspace State Management
+
+This part of the library is responsible for keeping track of the current workspace state. The internal workspace model is used to create requests for FCS. It needs to make sure the internal state doesn't get out of sync with what's in the editor and on disk.
+
+##### Things to keep track of:
+- Incoming document changes
+- Documents' open/closed status
+- Source files on-disk changes for documents that aren't open
+- Any on-disk references that could have been rebuilt/updated
+
+#### Project model
+
+We need some internal model for a workspace/solution with projects. We don't want to use any project system library because we don't need most of the functionality which will be instead handled by the client.
+
+The model needs to be able to quickly absorb incoming changes and easily convert to (or already contain) inputs for FCS. We could consider a reactive/adaptive approach for this.
+
+We should keep track of versions of each project and source file including all dependencies with regards to parsing and type checking.
+
+#### Example sequence diagram
+
+```mermaid
+
+sequenceDiagram
+
+participant Client
+participant FsLSP
+participant WorkspaceStateManagement
+participant FCS
+
+Client->>FsLSP: textDocument/didChange
+FsLSP->>WorkspaceStateManagement: Update workspace state
+Client->>+FsLSP: textDocument/diagnostic
+FsLSP->>+WorkspaceStateManagement: Get current workspace snapshot
+WorkspaceStateManagement-->>-FsLSP: Current workspace snapshot
+FsLSP->>+FCS: ParseAndCheckFileInProject
+FCS-->>-FsLSP: FSharpCheckFileResults
+FsLSP-->>-Client: DocumentDiagnosticReport
+
+```
+
+
+#### Concurrency and cancellation
+
+We need to make sure when a project snapshot is needed to process a request, it will always contain all change notifications that were received before. There shouldn't be need for a global request queue, possibly only for change notifications inside WorkspaceStateManagement. And any language requests will wait until WorkspaceStateManagement has absorbed all previous changes.
+
+For cancellation we should decide, potentially on a case-by-case basis what to do when an incoming change notification invalidates an ongoing request.
+
+1. Do nothing and let the request run either until it completes or is cancelled by the client. This is the easiest so we will probably start with this.
+2. Cancel the request and return "cancelled" result to the client. The client then needs to deal with it.
```

Comment:
IIRC FSAC does this automatic cancellation (at least of typechecking) and so this would be a take-back from users. I believe this matches 2+3 on your list here.

Reply:
In the end the invalidated type check (expensive operation) needs to be canceled. This is more about whose responsibility it will be and which protocol for that will give best performance. Will dig deeper into it when we're implementing it.